### ISSUE TITLE
Lack of context for failing reconciliation

### ISSUE THREAD

**Issue by tun0** [fluxcd/image-automation-controller#498]
```json
{
  "level": "error",
  "ts": "2023-03-20T14:08:31.180Z",
  "msg": "Reconciler error",
  "controller": "imageupdateautomation",
  "controllerGroup": "image.toolkit.fluxcd.io",
  "controllerKind": "ImageUpdateAutomation",
  "ImageUpdateAutomation": {
    "name": "apps",
    "namespace": "flux-system"
  },
  "namespace": "flux-system",
  "name": "apps",
  "reconcileID": "081a105f-7672-4fc7-b532-26be91972eeb",
  "error": "object not found"
}
```

This doesn't provide enough context to determine what actually is going wrong here.

**Comment by mantasaudickas** [comment]
I have the same error.. also no context and have no idea what's going on here :) 
Seems like everything is deployed, updated etc.. 
flux events does not show anything failing, and have no idea what I could describe with kubectl.. 
describe image-automation-controller does not show any new events at that time when error is displayed.

**Comment by kingdonb** [comment]
There are a couple of reports of this type of failure (or potentially unrelated failures eg. git error code 128) that are showing up in the Slack channel, I haven't seen them filter down to reports for IAC as of yet, but something to be aware of.

I will load up some Image Update Automation controls today or tomorrow and try to reproduce this issue, one or the other issue, there is not much context to go on for what is causing the failure. I understand this report is not about one specific failure, but the general case of failure not being reported very clearly with a good root cause.

```
{"level":"error","ts":"2023-04-03T19:27:38.561Z","msg":"Reconciler error","controller":"imageupdateautomation","controllerGroup":"image.toolkit.fluxcd.io","controllerKind":"ImageUpdateAutomation","ImageUpdateAutomation":{"name":"flux-system","namespace":"flux-system"},"namespace":"flux-system","name":"flux-system","reconcileID":"a43e903f-de19-4eaa-a7cd-e64a804d77fa","error":"malformed unpack status: \u00010069\u0001001dunpack index-pack failed\n0043ng refs/heads/main error processing packfiles: exit status 128\n0000"}
```

This is another example of that. This is the error returned from Git, and I'm not sure how much helpful parsing we can do, but to refocus, the subject of this report is about making it clearer what has gone wrong when IAC fails. Maybe we can come up with some common failure scenarios and start classifying errors to raise those as conditions, based on a pattern matching.

**Comment by mantasaudickas** [comment]
We have migrated our repository to another provider (migrated to gitlab from bitbucket.org). And seems like these errors are gone.
Nothing else changed: cluster and fluxcd versions remained the same. What I did just flux uninstall and flux bootstrap with new git repository url.
So seems like bitbucket.org have some specialties which produces this error?

**Comment by PaulSxxxs** [comment]
I'm having this same issue and have gone down the route of changing gitImplementation to use libgit2, but we're using source.toolkit.fluxcd.io/v1beta2 so this has been deprecated, (https://github.com/fluxcd/source-controller/blob/main/docs/spec/v1beta1/gitrepositories.md#git-implementation)

Furthermore v1beta2 recommends setting `--feature-gates=OptimizedGitClones=false` which I don't know how to achieve .. any tips on how to enable this? (https://github.com/fluxcd/source-controller/blob/main/docs/spec/v1beta2/gitrepositories.md#optimized-git-clones)

@mantasaudickas, We're also using bitbucket and have the scenario of multiple clusters using the same repo. Are you having good results so far?

**Comment by dewe** [comment]
> ... once FluxCD makes a push - its not possible to get that last push anymore (while it is visible using UI, but not fetchable to local copies and not visible in git command line history)... 

We've seen the same behaviour and have raised a support ticket with Bitbucket... still no solution though.

**Comment by mantasaudickas** [comment]
>have raised a support ticket with Bitbucket... still no solution though

Same here, since it was blocking us - we switched manifest repository location to another provider.. and now thinking to switch everything :)

**Comment by tobiasjochheimenglund** [comment]
Also getting the "Object not found" error using flux with Bitbucket. Imageautomation gets stuck starting with `image-automation-controller  command error on refs/heads/master: failed to update ref` . Manually pushing to the same repo seems to be a temporary fix

**Comment by PaulSxxxs** [comment]
@tobiasjochheimenglund We're seeing the same behavior's; image updater is returning `on-fast-forward update: refs/heads/master` while `--feature-gates=OptimizedGitClones=false`, but image updated so seem to be processing; I'm about to test this more thoroughly and will report back. For us too, the error disappears when someone makes a commit which is thankfully quite frequent.

@dewe @mantasaudickas Do you have any more technical details you sent to bitbucket to push the problem onto them if it does seem to be bitbucket specific?  Bitbucket didnt resolve for us either, though were helpful and pointed me towards git shallow clone potentially causing the issue. My support ticket was less technical and more a query about shallow clone, repo health and fluxcd. 

I'll report back with any findings.

**Comment by mantasaudickas** [comment]
They did not asked for any technical details.. all their communication sounded more like: please check that, or that and maybe we can do GC for your repo.. and I did not heard from them since last Friday :)
Not sure how shallow clone should cause such an issue - but sounds like just another: "we don't know what is happening and Git is not supposed to keep your history at all" :D

**Comment by youest** [comment]
Hello there, we have the same issue. our configuration is multicluster with different branches on the same Bitbucket repo.
We are encountering this error only on one branch/cluster but not in others, at least until now.
is there any configuration we can add or change to have more details to investigate the problem? for example, does it make sense to increase the log level?

**Comment by dewe** [comment]
@PaulSxxxs At the same time we get `object not found`, bitbucket pipeline doesn't trigger automatically as expected. When trying to start the pipeline manually it fails with "we couldn't clone the repository". There's definitely a correlation here. We have reported about the pipeline triggering problem, but got no actual response other than "Our engineering team is currently investigating this further".

**Comment by PaulSxxxs** [comment]
We had a similar issue committing from any git client for a time when "object not found" was occurring.
It definitely felt like some sort of lock but we couldn't figure it out, perhaps because it's a bitbucket issue.

**Comment by PaulSxxxs** [comment]
I received this message from Bitbucket:

`Syahrul commented:

G'day, Paul

A quick update on this issue.

Our development team noticed a pattern with the FluxCD issue with Bitbucket cloud. After thorough analysis, it has been determined that the issue is most likely caused by the go-git library being used by FluxCD. This library prematurely closes the connection before a push operation is completed.

To address this matter, we will release a fix tomorrow to mitigate the problem. Once the mitigation process is complete, we will provide you with an update.

We appreciate your patience and encourage you to reach out if you have any additional questions.

–
Mohammad Syahrul
Support Engineer APAC, Bitbucket cloud`

**Comment by mantasaudickas** [comment]
Wondering if "object not found" issue will be fixed, or its related to something else :)

**Comment by PaulSxxxs** [comment]
I specifically spoke to them about "object not found" and gave some technical details ... i'm fairly sure it will fix this.

**Comment by dewe** [comment]
Can't find any apparently related issue over at [go-git](https://github.com/go-git/go-git/issues)... 🤔

**Comment by hiddeco** [comment]
As I happen to be a go-git maintainer as well, we would be really happy to [see an issue being created in `go-git`](https://github.com/go-git/go-git/issues/new) with steps to reproduce (or any details they can share about how they determined the connection to be closed prematurely).

**Comment by gregawoods** [comment]
We use both flux and bitbucket and have been absolutely pulling our hair out over this issue. For what it's worth we found that moving from https:// to ssh:// git URLs seemed to make the behavior go away. That isn't always practical to do however so here's hoping that bitbucket's fix works out.

**Comment by PaulSxxxs** [comment]
Bitbucket rolled their fix, and for us everything has been working perfectly again.

**Comment by mantasaudickas** [comment]
Yeah... I reverted my manifests as well to bitbucket, so it works - but I am again getting "object not found" messages :)

**Comment by tun0** [comment]
Given that these `object not found` errors tend to get followed up by a successful reconciliation, it seems there's already some retry mechanism in place. Depending on the details of that retry logic, it might make sense to just "ignore" `object not found` error, unless it persists after several retries?

**Comment by pjbgf** [comment]
@tun0 Image Automation Controller would automatically retry on the next reconciliation, so yes it should be safe to disregard the "one-off" `object not found` error.

However, if you do find a pattern where you can reliably reproduce the issue, please report it [upstream](https://github.com/go-git/go-git/issues/new) so it can be investigated and fixed.

**Comment by mantasaudickas** [comment]
>However, if you do find a pattern where you can reliably reproduce the issue
It is still happening with Bitbucket Cloud :)

**Comment by hiddeco** [comment]
Then please report it upstream with more details around any patterns you observe while the error occurs (or e.g. information about the contents of your repository, size, etc.)?

There is little we can do from within the context of this repository, and it really has to be addressed there. Thanks for your cooperation.
